Fix #2992 of Creating new solution using "dotnet new sln" in .NET 5.0.201 SDK has wrong/invalid solution version number

eriawan opened the pull request:
### Fix #2992 of Creating new solution using "dotnet new sln" in .NET 5.0.201 SDK has wrong/invalid solution version number.

### Problem
I already detailed the problems in issue #2992

### Solution
Use valid solution version number for VS 2019 that always use this scheme of `16.0.xxxx.xxxx` instead of having invalid solution version number of `16.6.xxx.xx` as this breaks actual valid solution's minimum version. See also this comment: https://github.com/dotnet/templating/issues/2992#issuecomment-815281973
Please review. The detailed reasons why I submit this in issue #2992 

FYI, I'm using the version that reverted back to original solution version in the initial history of this file, based on this commit in this branch:  
https://github.com/dotnet/templating/commit/95d19ea6e0d10fc839f273223a736279464b3d38#diff-59d7a44760febde566cb20d6692d2a810281017b42d23841dd1e33eafa6275b1

![image](https://user-images.githubusercontent.com/8773147/114096717-4f4a3f80-98e9-11eb-9c71-2ae0dcc586b1.png)

### Checks:
- [ ] Added unit tests (_I think it is N/A as this is to fix to use correct solution's version number_)
- [ ] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

marcpopMSFT commented on template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln:
```diff
@@ -1,8 +1,8 @@
-VisualStudioVersion = 16.6.30114.105
-MinimumVisualStudioVersion = 10.0.40219.1
+# Visual Studio Version 15
+VisualStudioVersion = 15.0.26124.0
```
Is there a reason you went to 15 for VS Version rather than just fixing the 16 number to be 16.0.30114.105 which is the correct value for 16.6?

eriawan replied:
Thanks for the suggestion! 
I have just updated with the version in this commit https://github.com/dotnet/templating/pull/3007/commits/a971eee1f63af668b7d83bec27ebf01aaf0c3575

eriawan commented on template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln:
```diff
@@ -1,8 +1,8 @@
-MinimumVisualStudioVersion = 10.0.40219.1
+# Visual Studio Version 15
+VisualStudioVersion = 15.0.26124.0
+MinimumVisualStudioVersion = 15.0.26124.0
```
@marcpopMSFT 

Pardon, could you explain what is the technical reason of using version number of 10.0.xxx.xxx? What about the previous behavior in .NET Core 3.0 and 3.1 SDKs?
I mean, as far as I know, all of the solution generated by .NET Core 3.0.xxx, and also in Core 3.1.101 to 3.1.407 SDK always have minimum version set to `15.0.26124.0`, for example, this is the SLN generated from 3.1.406 SDK:

```
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio 15
VisualStudioVersion = 15.0.26124.0
MinimumVisualStudioVersion = 15.0.26124.0
Global
        GlobalSection(SolutionConfigurationPlatforms) = preSolution
                Debug|Any CPU = Debug|Any CPU
                Debug|x64 = Debug|x64
                Debug|x86 = Debug|x86
                Release|Any CPU = Release|Any CPU
                Release|x64 = Release|x64
                Release|x86 = Release|x86
        EndGlobalSection
        GlobalSection(SolutionProperties) = preSolution
                HideSolutionNode = FALSE
        EndGlobalSection
EndGlobal
```

But if this change of versioning behavior from previous Core 3.0 and 3.1 is intended and should be enforced, then I will change this to use version `10.0.40219.1` instead. Please confirm and thanks! 🙂

marcpopMSFT replied:
I'd try to have dotnet match what VS does which is target 10.  It's possible 3.0 picked 15 since .net core didn't exist back in 10.0.  Ultimately, I'll leave this to @donJoseLuis as the owner of the templating repo to make the call on what he wants the values to be.

eriawan replied:
@marcpopMSFT 

Ah, you're absolutely right! I have checked, and the `MinimumVisualStudioVersion` from VS 2019 is `10.0.40219.1`

I have updated this MinimumVisualStudioVersion` to use the 10.0.40219.1` version number.
Please review.

I also welcome @donJoseLuis to review 🙂

eriawan commented:
@marcpopMSFT 
I have to rebase it, because I mixed different git config between my local and Github direct edit.

Is there anything else besides your comment in https://github.com/dotnet/templating/pull/3007#discussion_r610119707 ?
Please confirm if this behavior changes are intended because the Core 3.0 and 3.1 always has `MinimumVisualStudioVersion ` set to other version higher than `10.0.xxxx.xxx`. when you are available.

Many thanks for your patience and review!

eriawan commented:
@marcpopMSFT 

I think I have addressed all of your feedback. 

cc @donJoseLuis 

If you have some time, please review and thanks in advance, guys! 🙂 🙏

eriawan commented:
hi @donJoseLuis 

Is there anything I should do further for this PR? Please review...

cc @marcpopMSFT

DavidKarlas reviewed (changes requested):
Hi, could you undo BOM change at beginning of file? I'm worried VS or some other tool might break without it which some unicode project/file names.

dsplaisted commented on template_feed/Microsoft.DotNet.Common.ItemTemplates/content/Solution/Solution1.sln:
```diff
@@ -1,7 +1,7 @@
-﻿
+
```
Apparently there are some whitespace changes here.  I don't know if that will affect anything but I think they should be reverted to be safe.

eriawan replied:
@dsplaisted 
Thanks!
Done, whitespace changes reverted.

eriawan commented:
@DavidKarlas 

Thanks for the catch! My bad!
I have reverted to have correct BOM.  Please review, thanks in advance!

eriawan commented:
@dsplaisted @DavidKarlas 

I think I have addressed all of the remaining PR feedback.
Please review and let me know if you have other feedback. 
Thanks in advance! 🙂 🙏

eriawan commented:
@dsplaisted @DavidKarlas 

Pardon I keep bugging you, folks...
Now that this PR has been approved, is there anything else I have to do/edit? ❓ 
Because I'd love to port this PR to branches of `release/5.0.1xx` and `release\5.0.3xx` if this got merged. 🙏 

Thanks in advance!

eriawan commented:
folks,

Please let me know, is this PR needs further triage/discussion for the team, or should I add more/update more? 
Since this PR has been approved, do you need more time to discuss with the rest of .NET Core team? Or more approvals?
Please confirm. It's been 3 days with no further update.. 

cc @dsplaisted @DavidKarlas

DavidKarlas commented:
Hi @eriawan,

problem is that this PR is targeting `2xx` branch and not `3xx` or `main`, if you open PR against `3xx` and `main` we can merge that right away.

Merging things  into `2xx` is much harder. Since we need to argument why change meets bar for servicing and honestly, I don't think this bug is big enough to justify servicing and I wanted input from others before bringing it up.

I think that modifying .sln by hand is super edge case scenario and if you are advanced user enough to modify .sln by hand you can probably figure out what versions to use.

Sorry for not responding sooner.

Regards,
David

eriawan commented:
@DavidKarlas 

Many thanks for your update!
About this:

> problem is that this PR is targeting 2xx branch and not 3xx or main, if you open PR against 3xx and main we can merge that right away.

Could I just directly change this PR's target branch to `release/5,0.2xx` to 'release/5.0.3xx` or do you want me to close this and create new 2 PR target 'release/5.0.3xx` branch and `main` branch separately?

Please confirm.

DavidKarlas commented:
Preferably change this PR to `main` so once its merged, in future people can get back to this PR with all the context.
And open new one against `3xx`.

eriawan commented:
@DavidKarlas 

done changing target branch to `main`, and now I have merge conflicts.
going to resolve these first...

eriawan commented:
@DavidKarlas 

I have updated base branch to main as you have suggested, but then I got too many conflicting files...
since the conflicts are too many for me to quickly resolve, could I open a new PR with target branch to `main` instead?

eriawan commented:
@DavidKarlas 

Are you currently busy now? I mean, please advice, as I have changed the target branch to `main` as suggested then I have conflicting files.

I still think it's easier to just create new PR that target `main` branch, instead of trying to resolve these numerous conflicting files. And yes, I agree that the context will not be lost if I just change the target branch. It's now I don't know how to qiuckly resolve these conflicting files from github web interface directly instead of resolving on my machine one by one.

Do you want me to solve this one by one for each conflicting files now? This will take some time, and I don't think I could finish this within 5 to 10 minutes, so I may revisit this later if I have enough time... 

Please let me know what do you think/suggest about this. Or, if you are busy, please let me know soon so I could revisit this PR in other time when you are available, while I continue to use 3.1.xxx instead for my team and my customers.

DavidKarlas commented:
Its OK to close this PR and open new one, I forgot it would create all this conflicts, sorry about that...
We can just mention this PR in other PRs and it will be fine!

eriawan commented:
@DavidKarlas 

Many thanks!
New PR to target `main` branch has been submitted: #3026 
I'm closing this PR now.

DavidKarlas commented:
Can you also open for `3xx`?

eriawan commented:
@DavidKarlas 

> Can you also open for 3xx?

Sure! I'm going to submit now

eriawan commented:
I'm going to retarget this PR to 5.0.2xx in case this PR is approved to be included as servicing.

eriawan commented:
Closing this PR after changing target branch back to 5.0.2xx.

DavidKarlas commented:
Unfortunately this doesn't meet servicing bar but since its merged into 5.0.3xx branch, it will be available on stable channel in May which would also be true if it was merged into 5.0.2xx...

eriawan commented:
@DavidKarlas 

Thanks for your reply!

> it will be available on stable channel in May which would also be true if it was merged into 5.0.2xx

So 5.0.3xx will be available on May? what do you mean "_**would also be true if it was merged into 5.0.2xx**_" ?

DavidKarlas commented:
I meant to say that next 2xx version will also be released in May.
